src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -294,12 +280,20 @@ public static Task<CancellationToken> FromTimeoutAsync(int millisecondsTimeout)
                 {
                     // Since this thread was successful reserving a spot in the cache, it would be the only thread
                     // that construct the CancellationTokenSource
-                    var token = new CancellationTokenSource((int)(targetTime - currentTime)).Token;
+                    var tokenSource = new CancellationTokenSource((int)(targetTime - currentTime));
+                    var token = tokenSource.Token;
 
                     // Clean up cache when Token is canceled
                     token.Register(t => {
                         Task<CancellationToken> ignored;
-                        s_tokenCache.TryRemove((long)t, out ignored);
+                        try
+                        {
+                            s_tokenCache.TryRemove((long)t, out ignored);
+                        }
+                        finally
+                        {
+                            tokenSource.Dispose();
```

Comment:
This will cause a closure to be created as you are capturing the `tokenSource` local variable. After the timer has fired (which will have happened when you're getting into this code block), is there any work needed to be done by `Dispose()`? I believe the timer is a fire once timer so it will no longer be registered and there's nothing left in the timer to clean up on Disposal. If there is some work which is beneficial to be done here, one option would be to change the type of `s_tokenCache`  to `ConcurrentDictionary<long, Task<CancellationTokenSource>>`. This would make the type of ignore to be `Task<CancellationTokenSource>` and it would no longer be ignored :smile: